include/pybind11/stl/filesystem.h

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2021 The Pybind Development Team.
+// All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+#pragma once
+
+#include "../detail/common.h"
+
+#ifdef __has_include
+#  if defined(PYBIND11_CPP17) && __has_include(<filesystem>) && \
+      PY_VERSION_HEX >= 0x03060000
+#    include <filesystem>
+#    define PYBIND11_HAS_FILESYSTEM 1
+#  endif
+#endif
```

Comment:
Just a thought, do we need this `PYBIND11_HAS_FILESYSTEM` now? If I include `<stl/filesystem.h>` do I want it to silently not cast things instead of getting a compile-time error?

Reply:
(If not, it should be moved to the include in the test ;) )

Reply:
> (If not, it should be moved to the include in the test ;) )

Copying my question from the slack channel here:

> In this first pass, I simply kept the #ifdef block defining PYBIND11_HAS_FILESYSTEM in the header file. Does that still make sense? Would it be better to move that to the test? — It seem useful to keep it in the header, so others can easily use it if available without having to do anything much otherwise, but I'm not sure how this shakes out best.

Ideally we'd give people a choice? Bail out with a compile-time error or have them handle the runtime error.

It's not really great if we lead users down the path of copy-pasting the `__has_include` logic from the test.

Reply:
But do users want "works if available" behavior for this? (The answer may be yes) It seems a little weird if your package only accepts `pathlib.Path` arguments if it was compiled in C++17 mode on some compilers. And only allows you to return std::filesystem on Python 3 (though that I'm less worried about).  I'm guessing there will be some extra work most users will have to do to make sure this works (including requiring macOS 10.15+), so forcing the user to explicit might be better.

That is, I expect most users will either make an effort to support std::filesystem/pathlib everywhere, or they will not support it, but very few will support it if available. But could easily be wrong. I don't find the "auto" selection of C++ level all that useful either, but some users want it.

Reply:
I'm okay to leave it as is, a user (who looks at the source) can check `PYBIND11_HAS_FILESYSTEM` if they want. But I'm also okay to (re)move it to the tests.

Reply:
I added `PYBIND11_HAS_FILESYSTEM_IS_OPTIONAL`.
If not defined, users will see a compile-time error.
If that's not desired, they can opt in to handling the runtime error, without having to copy-paste the `__has_include` logic.

I don't think we need to document this. The safe behavior (compile-time error) is the default, the error message tells people what the problem is and what they can do to ignore it.

Reply:
`__has_include` is not actually a C++17 feature, it's a [C++20 one]()https://en.cppreference.com/w/cpp/feature_test. Though I think all major compilers added it very early, but still, maybe it's best to remove this and just `#include` filesystem and let the failure happen if it's not available, rather than calling `#error`. This also makes `PYBIND11_HAS_FILESYSTEM` rather weird, though.

Reply:
@henryiii This page suggest it is a C++17 feature though? https://en.cppreference.com/w/cpp/preprocessor/include

Reply:
official spec: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/p0061r1.html

Reply:
`__has_include` currently appears in 10 source code locations on master (see below).

If think for this PR it's important to stay focused on the one problem we want to solve as soon as possible: users shouldn't be forced to work on the link options if they don't actually want to use the new caster. I'll update the PR description accordingly.

```
include/pybind11/cast.h:#  if defined(__has_include)
include/pybind11/cast.h:#    if __has_include(<string_view>)
include/pybind11/stl.h:#ifdef __has_include
include/pybind11/stl.h:#  if defined(PYBIND11_CPP17) && __has_include(<optional>)
include/pybind11/stl.h:#  if defined(PYBIND11_CPP14) && (__has_include(<experimental/optional>) && \
include/pybind11/stl.h:                                 !__has_include(<optional>))
include/pybind11/stl.h:#  if defined(PYBIND11_CPP17) && __has_include(<variant>)
include/pybind11/stl.h:#  if defined(PYBIND11_CPP17) && __has_include(<filesystem>) && \
include/pybind11/detail/common.h:#if defined(__has_include)
include/pybind11/detail/common.h:#  if __has_include(<version>)
```

Reply:
Yes, @Skylion007 is correct, I was muddling has_include with feature test macros - probably shouldn't comment on things while in a conference. :)

This plans seems fine, we need to fix this, then we can polish and or document a bit more later if needed.